src/consensus/blockstore.rs

```diff
@@ -105,230 +81,121 @@ impl Blockstore {
-                .insert((slot, slice), shred.merkle_root);
-        } else if cached_merkle_root != Some(&shred.merkle_root) {
-            if self.equivocated_slots.insert(slot) {
-                warn!("shreds show leader equivocation in slot {slot}");
+        let leader_pk = self.epoch_info.leader(slot).pubkey;
+        let event = self
+            .slot_data_mut(slot)
+            .add_shred(shred.clone(), check_equivocation, leader_pk)
+            .await;
+
+        match event.clone()? {
+            VotorEvent::FirstShred(_) => {
+                // notify Votor of first slice
+                self.votor_channel.send(event.unwrap()).await.unwrap();
+                None
             }
-            if check_equivocation {
-                return None;
+            VotorEvent::Block { slot, block_info } => {
+                // notify Votor of block and print block info
+                self.votor_channel.send(event.unwrap()).await.unwrap();
+                debug!(
+                    "reconstructed block {} in slot {} with parent {} in slot {}",
+                    &hex::encode(block_info.hash)[..8],
+                    slot,
+                    &hex::encode(block_info.parent_hash)[..8],
+                    block_info.parent_slot,
+                );
+
+                Some((slot, block_info))
             }
-        }
-
-        // store and handle this shred if and only if:
-        //   - it is not yet stored in the blockstore
-        //   - it is not (known to be) after the last slice
-        let exists = slice_shreds.iter().any(|s| {
-            s.payload().index_in_slice == index
-                && shred.payload_type.is_data() == s.payload_type.is_data()
-        });
-        let after_last = self
-            .last_slices
-            .get(&slot)
-            .is_some_and(|last_slice| slice > *last_slice);
-        if exists || after_last {
-            return None;
-        }
-        slice_shreds.push(shred);
-
-        // store last slice index, delete later shreds
-        if is_last_slice && !self.last_slices.contains_key(&slot) {
-            self.last_slices.insert(slot, slice);
-
-            // delete shreds after last slice, if there are any
-            let keys_to_delete: Vec<_> = self
-                .shreds
-                .range(&(slot, slice + 1)..&(slot + 1, 0))
-                .map(|(k, _)| k)
-                .copied()
-                .collect();
-            for key in keys_to_delete {
-                self.shreds.remove(&key);
-            }
-        }
-
-        // maybe send first shred notification
-        if self.first_shred_seen.insert(slot) {
-            let event = VotorEvent::FirstShred(slot);
-            self.votor_channel.send(event).await.unwrap();
-        }
-
-        // maybe reconstruct slice and block
-        if self.try_reconstruct_slice(slot, slice) {
-            self.try_reconstruct_block(slot).await
-        } else {
-            None
+            _ => unreachable!(),
```

Comment:
```suggestion
            ev => panic!("unexpected event {ev:?}"),
```

Will help with debugging in future.